streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCacheBytes.java

```diff
@@ -0,0 +1,442 @@
+        this.maxCacheSizeBytes = maxCacheSizeBytes;
+        this.map = new HashMap<>();
+        listeners = new HashMap<>();
+        dirtyKeys = new HashMap<>();
+    }
+
+    /**
+     * Add a listener that is called each time an entry is evicted from the cache or an explicit flush is called
+     * @param namespace
+     * @param listener
+     */
+    public synchronized void addDirtyEntryFlushListener(final String namespace, DirtyEntryFlushListener listener) {
+        this.listeners.put(namespace, listener);
+    }
+
+    public synchronized void flush(final String namespace) {
+        final DirtyEntryFlushListener listener = listeners.get(namespace);
+        if (listener == null) {
+            throw new IllegalArgumentException("No listener for namespace " + namespace + " registered with cache");
+        }
+
+        final Set<Bytes> keys = dirtyKeys.get(namespace);
+        if (keys == null || keys.isEmpty()) {
+            return;
+        }
+        final List<DirtyEntry> entries  = new ArrayList<>();
+        for (Bytes key : keys) {
+            final MemoryLRUCacheBytesEntry entry = get(namespace, key.get());
+            entries.add(new DirtyEntry(key, entry.value, entry));
+        }
+        listener.apply(entries);
+        keys.clear();
+    }
+
+    private void callListener(final Bytes key, final LRUNode node) {
+        final MemoryLRUCacheBytesEntry entry = node.entry();
+        final DirtyEntryFlushListener listener = listeners.get(node.namespace);
+        if (listener == null) {
+            return;
+        }
+        listener.apply(Collections.singletonList(new DirtyEntry(key, entry.value, entry)));
+    }
+
+    public synchronized MemoryLRUCacheBytesEntry get(final String namespace, byte[] key) {
+        MemoryLRUCacheBytesEntry entry = null;
+        // get map
+        final TreeMap<Bytes, LRUNode> treeMap = this.map.get(namespace);
+        if (treeMap != null) {
+            // get element
+            LRUNode node = treeMap.get(cacheKey(key));
+            if (node != null) {
+                entry = node.entry();
+                updateLRU(node);
+            }
+        }
+        return entry;
+    }
+
+    private void maybeEvict() {
+        while (sizeBytes() > maxCacheSizeBytes) {
+            final Bytes key = tail.key;
+            final String namespace = tail.namespace;
+            final TreeMap<Bytes, LRUNode> treeMap = this.map.get(namespace);
+            final LRUNode toRemove = treeMap.get(key);
+            currentSizeBytes -= toRemove.size();
+            if (toRemove.entry.isDirty()) {
+                callListener(toRemove.key, toRemove);
+            }
+            remove(toRemove);
+            treeMap.remove(key);
+
+            if (toRemove.entry.isDirty()) {
+                removeDirtyKey(namespace, key);
+            }
+        }
+    }
+
+
+    public synchronized void put(final String namespace, byte[] key, MemoryLRUCacheBytesEntry value) {
```

Comment:
Actually I'm thinking we we should have a per-thread cache with size divided by number of threads in order to avoid concurrent access? In practice we recommend one thread per client, so that `ThreadCache` is actually the whole cache.